solr/core/src/java/org/apache/solr/cloud/ZkController.java

```diff
@@ -2832,9 +2841,28 @@ public boolean checkIfCoreNodeNameAlreadyExists(CoreDescriptor dcore) {
    */
   public void publishNodeAsDown(String nodeName) {
     log.info("Publish node={} as DOWN", nodeName);
-    ZkNodeProps m = new ZkNodeProps(Overseer.QUEUE_OPERATION, OverseerAction.DOWNNODE.toLower(),
-        ZkStateReader.NODE_NAME_PROP, nodeName);
     try {
+      Set<String> processedCollections = new HashSet<>();
```

Comment:
Looks like there is no use of `processedCollections`, we can remove now?

Reply:
It is used on line 2848: https://github.com/fullstorydev/lucene-solr/pull/183/files/f0193f630ec2faa0d218216aed7ea93e322cde81#diff-5b63503605ede4384429e74d1fa0c410adc5da8f3246e8c36e49feff2f3ea692R2848 to track if we've already done that collection. If `processedCollections` contains the collection name then the add method will return false and it won't be processed again.

Reply:
Ah, thanks.

Reply:
Sorry for weighing in belatedly.

Is there a problem if a collection gets processed twice? I ask because `parallelStream()` results in entries being processed concurrently, and `processedCollections` is not a thread-safe map.

I think `cc.getCoreDescriptors()` ends up being an ArrayList, which should indeed be eligible to be processed concurrently by `parallelStream()`

I see three options:
1. if double-processing a collection is fine, it's just a "best-effort" thing -- practically speaking we could probably leave as-is.
2. switch to `stream()` if we don't mind (or would prefer) to process serially.
3. make processedCollections a `ConcurrentHashMap.newKeySet()`

My gut sense is that probably 3. would be the best choice here?

Reply:
Good catch, I agree with 3 and I can make that update. I think the cost of processing a collection twice is fairly low, just updating the same ZK files again, but it seems easy enough to go for the better implementation.